DSharpPlus/Entities/Channel/DiscordChannel.cs

```diff
@@ -624,34 +624,105 @@ public async Task<ThreadQueryResult> ListJoinedPrivateArchivedThreadsAsync(DateT
     }
 
     /// <summary>
-    /// Deletes multiple messages if they are less than 14 days old.  If they are older, none of the messages will be deleted and you will receive a <see cref="BadRequestException"/> error.
+    /// Deletes multiple messages if they are less than 14 days old.  If they are older, none of the messages will be deleted and you will receive a <see cref="ArgumentException"/> error.
     /// </summary>
     /// <param name="messages">A collection of messages to delete.</param>
     /// <param name="reason">Reason for audit logs.</param>
-    /// <returns></returns>
-    /// <exception cref="UnauthorizedException">Thrown when the client does not have the <see cref="Permissions.ManageMessages"/> permission.</exception>
-    /// <exception cref="NotFoundException">Thrown when the channel does not exist.</exception>
-    /// <exception cref="BadRequestException">Thrown when an invalid parameter was provided.</exception>
-    /// <exception cref="ServerErrorException">Thrown when Discord is unable to process the request.</exception>
-    public async Task DeleteMessagesAsync(IEnumerable<DiscordMessage> messages, string reason = null)
+    /// <returns>The number of deleted messages</returns>
+    /// <remarks>One api call per 100 messages</remarks>
+    public async Task<int> DeleteMessagesAsync(IReadOnlyList<DiscordMessage> messages, string? reason = null)
     {
-        // don't enumerate more than once
-        ulong[] msgs = messages.Where(x => x.Channel.Id == this.Id).Select(x => x.Id).ToArray();
-        if (messages == null || !msgs.Any())
+        ArgumentNullException.ThrowIfNull(messages, nameof(messages));
+        int count = messages.Count;
+
+        if (count == 0)
         {
             throw new ArgumentException("You need to specify at least one message to delete.");
         }
+        else if (count == 1)
+        {
+            await this.Discord.ApiClient.DeleteMessageAsync(this.Id, messages[0].Id, reason);
+            return 1;
+        }
+        
+        int deleteCount = 0;
 
-        if (msgs.Count() < 2)
+        try
+        {
+            for (int i = 0; i < count; i += 100)
+            {
+                int takeCount = Math.Min(100, count - i);
+                DiscordMessage[] messageBatch = messages.Skip(i).Take(takeCount).ToArray();
```

Comment:
Unnecessary ToArray call. You should be able to use a range here instead.

Reply:
You're sure? the foreach uses it and we do a Select later on.

Reply:
Yeah, the range call should be much cheaper than the ToArray call. Something like
```cs
foreach(DiscordMessage message in messages[i..(i + takeCount)])
{
    // ...
    await this.Discord.ApiClient.DeleteMessagesAsync(this.Id, messages[i..(i + takeCount)].Select(x => x.Id), reason);
}
```

Reply:
That's exactly the one we had issues with before

Reply:
Testing shows it should work as expected.

![image](https://github.com/DSharpPlus/DSharpPlus/assets/46751150/45922954-b935-4a2e-a9fb-9dd45fba46d4)